packages/compartment-mapper/src/node-modules.js

```diff
@@ -252,24 +252,49 @@ const graphPackage = async (
   graph[packageLocation] = /** @type {Node} */ (result);
 
   /** @type {Record<string, string>} */
-  const dependencies = {};
+  const dependenciesObj = {};
   const children = [];
-  const predicates = packageDescriptor.dependencies || {};
+  const optionals = new Set();
+  const {
+    dependencies = {},
+    peerDependencies = {},
+    peerDependenciesMeta = {},
+    bundleDependencies = {},
+    optionalDependencies = {},
+    devDependencies = {},
+  } = packageDescriptor;
+  const allDependencies = {};
+  assign(allDependencies, dependencies);
+  assign(allDependencies, peerDependencies);
+  Object.keys(peerDependenciesMeta).forEach(name => {
+    if (peerDependenciesMeta[name].optional) {
+      optionals.add(name);
+    }
+  });
```

Comment:
Please consider, in keeping with the local rustic æsthetic:

```suggestion
  for (const [name, { optional }] of Object.entries(peerDependenciesMeta)) {
    if (optional) {
      optionals.add(name);
    }
  }
```

If we went more toward the array functional style, I’d be suggesting `entries`, `filter` and `forEach(set.add, set)`, but my prevailing preference is to use loops.

The contents of `package.json` are user data to us. We probably should be validating the package descriptor with something similar to what we use to validate compartment maps, before we even see it here. I don’t think we’re validating package descriptors at all.

Filed https://github.com/endojs/endo/issues/1357